### ISSUE TITLE
[FSDP] Customizable gradient pre-divide for mixed precision training

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#78774]
### 🚀 The feature, motivation and pitch

When training FSDP with mixed precision and fp16, we need to be aware of the reduced range and avoid over/underflow. If we sum all gradients for a large world size and then divide, this might overflow the fp16 range and cause issues, while if we divide first and then sum, we may risk underflow. One possible solution is to divide by sqrt(N), run allreduce, and then divide the result again by sqrt(N) which is equivalent to a N = world size divide. As a result, allreduce only increases the magnitude by sqrt(N=world size), which is less likely to cause overflow.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by kwen2501** [comment]
Check this NCCL API: `ncclRedOpCreatePreMulSum`
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/ops.html

It is a more versatile version than `ncclAvg` (which uses N as the pre-divider). `ncclRedOpCreatePreMulSum` allows using any scalar as the pre-divider.

**Comment by mori360** [comment]
We are committed to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
